File: 0073 Set Matrix Zeroes/set_matrix_zeroes.cpp
/*
BY RUSHABH GALA
rushabhvg
GitHub :
https://github.com/rushabhvg
StackOverflow :
https://stackoverflow.com/users/16571212/rushabhvg
LinkedIn:
https://www.linkedin.com/in/rushabhvg/
*/

class Solution {
public:
    void setZeroes(vector<vector<int>>& matrix) {
        int n = matrix.size();
        int m = matrix[0].size();
        bool f1 = false, f2 = false;
        for(int i=0; i<n; i++) {
            for(int j=0; j<m; j++) {
                if(matrix[i][j] == 0){
                    if(i==0) f1=true;
                    if(j==0) f2=true;
                    matrix[i][0] = 0;
                    matrix[0][j] = 0;
                }
            }
        }
        for(int i=1; i<n; i++) {
            for(int j=1; j<m; j++) {
                if(matrix[i][0] == 0 || matrix[0][j] == 0){
                    matrix[i][j] = 0;
                }
            }
        }
        if(f1) for(int i=0; i<m; i++) matrix[0][i]=0;
        if(f2) for(int i=0; i<n; i++) matrix[i][0]=0;
    }
};

/*
BY RUSHABH GALA
rushabhvg
GitHub :
https://github.com/rushabhvg
StackOverflow :
https://stackoverflow.com/users/16571212/rushabhvg
LinkedIn:
https://www.linkedin.com/in/rushabhvg/
*/

File: 73-set-matrix-zeroes/set-matrix-zeroes.java
/*
By Rushabh Gala
rushabhvg
GitHub:
https://github.com/rushabhvg
StackOverflow:
https://stackoverflow.com/users/16571212/rushabhvg
LinkedIn:
https://www.linkedin.com/in/rushabhvg/
*/

class Solution {
    public void setZeroes(int[][] matrix) {
        int n = matrix.length;
        int m = matrix[0].length;
        int[] col = new int[m];
        int[] row = new int[n];
        for(int i=0; i<n; i++) {
            for(int j=0; j<m; j++) {
                if(matrix[i][j]==0) {
                    row[i] = 1;
                    col[j] = 1;
                }
            }
        }
        // int k;
        // for(k=0; k<Math.min(m, n); k++) {
        //     System.out.print(k +" => " + row[k] + "\t" + col[k]+ "\n");
        // }
        // while(k<n) {
        //     System.out.print(k +" => " + row[k++] + "\n");
        // }
        // while(k<m) {
        //     System.out.print(k +" => \t" + col[k++] + "\n");
        // }
        for(int i=0; i<n; i++) {
            if(row[i]==1) {
                for(int j=0; j<m; j++) {
                    matrix[i][j] = 0;
                }
            }
        }
        for(int i=0; i<n; i++) {
            for(int j=0; j<m; j++) {
                if(col[j]==1) {
                    matrix[i][j] = 0;
                }
            }
        }
        return ;
    }
}

/*
By Rushabh Gala
rushabhvg
GitHub:
https://github.com/rushabhvg
StackOverflow:
https://stackoverflow.com/users/16571212/rushabhvg
LinkedIn:
https://www.linkedin.com/in/rushabhvg/
*/
